src/clients/python/simple_callback_client.py

```diff
@@ -0,0 +1,157 @@
+            self._input0_data = input0_data
+            self._input1_data = input1_data
+            self._request_cnt = 2
+            self._done_cnt = 0
+            self._recent_ctx_obj = None
+            self._recent_request_id = None
+            self._success = True
+    
+    context = Context(input0_data, input1_data)
+    def active_callback(context, idx, ctx_obj, request_id):
+        result = ctx_obj.get_async_run_results(request_id, True)
+        context._cv.acquire(True)
+        print("Callback " + str(idx) + " is called")
+        output0_data = result['OUTPUT0'][0]
+        output1_data = result['OUTPUT1'][0]
+        print(output0_data)
+        print(output1_data)
+        for i in range(16):
+            if (context._input0_data[i] + context._input1_data[i]) != output0_data[i]:
+                print("error: incorrect sum")
+                context._success = False
+                break
+            if (context._input0_data[i] - context._input1_data[i]) != output1_data[i]:
+                print("error: incorrect difference")
+                context._success = False
+                break
+        context._done_cnt += 1
+        context._cv.notify_all()
+        context._cv.release()
+
+    # Send async inference and wait for them to finish
+    for idx in range(context._request_cnt):
+        result = ctx.async_run_with_cb(partial(active_callback, context, idx),
+                                  { 'INPUT0' : (input0_data,),
+                                    'INPUT1' : (input1_data,) },
+                                  { 'OUTPUT0' : InferContext.ResultFormat.RAW,
+                                    'OUTPUT1' : InferContext.ResultFormat.RAW },
+                                  batch_size)
+
+    context._cv.acquire()
+    while context._done_cnt != context._request_cnt:
+        context._cv.wait()
+    context._cv.release()
+
+    if not context._success:
+      sys.exit(1)
+
+    # defer retrieval to main thread
+    def passive_callback(context, ctx_obj, request_id):
+        context._cv.acquire(True)
+        # set related InferContext and request id to shared Context object
+        context._recent_ctx_obj = ctx_obj
+        context._recent_request_id = request_id
+        context._cv.notify_all()
+        context._cv.release()
+
+    # Send async inference and wait for its callback is invoked
+    result = ctx.async_run_with_cb(partial(passive_callback, context),
+                              { 'INPUT0' : (input0_data,),
+                                'INPUT1' : (input1_data,) },
+                              { 'OUTPUT0' : InferContext.ResultFormat.RAW,
+                                'OUTPUT1' : InferContext.ResultFormat.RAW },
+                              batch_size)
+
+    context._cv.acquire()
+    while context._recent_ctx_obj is None:
+        context._cv.wait()
+    context._cv.release()
+
+    # Retrieve InferContext and request id from the shared context
```

Comment:
Lots of common code with above that should be combined into a "checking" thread or future.

Reply:
@deadeyegoodwin I simplified the callback example. Although GtiHub is still showing "change requested"